Adding logging for Hydrography scripts

varun-andhra-mapup opened the pull request.

maneesh-mapup commented on hydrography-approach/run-hydrography-pipeline.py:
```diff
@@ -35,11 +44,11 @@ def main():
 
     # Load config file
-    print("\nLoading the config file.")
+    logging.info("Loading the config file.")
```
@varun-andhra-mapup Looks like you did not provide a response to AI comments. Please do before asking me to merge thie

varun-andhra-mapup replied:
I had raised PR to receive an initial PR review. I have made commits based on feedback and added comments as well.

maneesh-mapup commented on hydrography-approach/run-hydrography-pipeline.py:
```diff
@@ -13,146 +15,204 @@
 
-    print("\nFiltering OSM ways data.")
-    filter_osm_ways.filter_ways(input_osm_pbf, output_osm_pbf, output_gpkg)
+    logger.info("Filtering OSM ways data.")
```
@varun-andhra-mapup can you refer to our meeting again to see how to close requests from PR agent review

varun-andhra-mapup replied:
I have made changes based on PR feedback and posted change comments as well.

varun-andhra-mapup commented:
### Fixed logging configuration
Updated the logging configuration to use a path specified in the `config.yml` file, replacing the hardcoded path. This enhances flexibility and ensures consistent logging across different environments.

### Commit:
https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/31/commits/e05fbc24cbc65fba561110d55895daa4aa7c9a98

### Code Improvement
```python
def main() -> None:
    ...
    # Configure logging after loading config
    log_file_path = config["logging"].get("log_file_path", "hydrography-pipeline.log")
    logging.basicConfig(
        filename=log_file_path,  # Use the path from the configuration
        level=logging.INFO,
        format="%(asctime)s - [%(levelname)s] - (%(filename)s).%(funcName)s - %(message)s",
    )
    ...

varun-andhra-mapup commented:
### Error Handling for Layer Loading

Improved error handling for layer loading by adding detailed logging and maintaining process integrity. The function now logs specific errors if layers fail to load and exits immediately, as retries are not practical given the likely file path issues. Continuing without these layers is not possible since subsequent processes depend on their successful loading.

### Commit
https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/31/commits/16da213f3fcda1c11261b3054ab8604c9aff9c98

### Code Snippet:
```python
from typing import Optional, Tuple
from qgis.core import QgsVectorLayer

def load_layers(nbi_points_fp: str, osm_fp: str, logger) -> Tuple[Optional[QgsVectorLayer], Optional[QgsVectorLayer]]:
    """
    Load required layers with improved error handling.

    Args:
        nbi_points_fp (str): File path to the NBI points layer.
        osm_fp (str): File path to the OSM ways layer.
        logger: Logger instance for logging errors.

    Returns:
        Tuple[Optional[QgsVectorLayer], Optional[QgsVectorLayer]]:
            A tuple containing the NBI points layer and the OSM ways layer.
    """
    def load_layer(fp: str, layer_name: str) -> Optional[QgsVectorLayer]:
        """
        Load a layer and log errors if the loading fails.

        Args:
            fp (str): File path to the layer.
            layer_name (str): Name of the layer for logging purposes.

        Returns:
            Optional[QgsVectorLayer]: Loaded layer or None if failed.
        """
        layer = QgsVectorLayer(fp, layer_name, "ogr")
        if not layer.isValid():
            logger.error(f"{layer_name} layer failed to load. Check the file path and ensure the file exists.")
            return None
        return layer

    nbi_points_gl = load_layer(nbi_points_fp, "nbi-points")
    osm_gl = load_layer(osm_fp, "filtered")

    if nbi_points_gl is None:
        logger.error("NBI points layer is critical and could not be loaded. Exiting.")
        sys.exit(1)

    if osm_gl is None:
        logger.error("OSM ways layer could not be loaded. Exiting.")
        sys.exit(1)

    return nbi_points_gl, osm_gl

varun-andhra-mapup commented:
### Improved Error Handling for CSV Writing**
Enhanced the robustness of the CSV file creation process by adding error handling for potential I/O errors. Previously, if there was an issue writing the CSV file, the application would fail silently without providing any indication of the failure.

### Commit
https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/31/commits/16da213f3fcda1c11261b3054ab8604c9aff9c98

**Changes Made:**
- Added a `try-except` block around the `df.to_csv` method to catch and log `IOError` exceptions.

```python
try:
    df.to_csv(intermediate_association)
    logger.info(f"{intermediate_association} file has been created successfully!")
except IOError as e:
    logger.error(f"Failed to write {intermediate_association}: {str(e)}")

varun-andhra-mapup commented:
### Ensure directory existence before use

This change adds a check to see if directories already exist before attempting to create them. It also includes logging for both successful directory creation and pre-existing directories.

### Code Improvement

- Implemented `os.path.exists()` to check directory existence.
- Added logging for directory creation and pre-existing directories.
- Improved error handling with try-except blocks.
```python
def create_directories(config: Dict[str, str]) -> None:
    """
    Create directories for output data as specified in the configuration.
    
    :param config: Dictionary containing paths for directories to be created.
    """
    output_folders = config.get("output_data_folders", {})

    # Define required folder keys
    required_folders = ["state_folder", "csv_files", "gpkg_files", "pbf_files"]

    for folder in required_folders:
        folder_path = output_folders.get(folder)
        if folder_path:
            if not os.path.exists(folder_path):
                try:
                    os.makedirs(folder_path)
                    logger.info(f"Directory created: {folder_path}")
                except Exception as e:
                    logger.error(f"Failed to create directory {folder_path}: {e}")
            else:
                logger.info(f"Directory already exists: {folder_path}")
        else:
            logger.warning(f"Path for {folder} not specified in configuration.")
```

This ensures directories are only created if they do not already exist and provides better feedback during execution.

varun-andhra-mapup commented:
/review

varun-andhra-mapup commented:
### Improved Error Handling in `load_layers`

The `load_layers` function now incorporates a custom `LayerLoadError` exception to handle errors more gracefully. This change replaces the use of `sys.exit(1)` with exceptions that can be caught and managed at a higher level, offering better control over the application flow.

### Commit
https://github.com/mapup/edit-osm-add-missing-bridge-for-truck-restriction/pull/31/commits/d8fc364f19e395175dc211d2d17e38f353afb956

### Code Improvement
```python
class LayerLoadError(Exception):
    """Custom exception for errors in loading vector layers."""

    def __init__(self, layer_name: str, message: str):
        super().__init__(f"{layer_name} layer: {message}")
        self.layer_name = layer_name


def load_layers(
    nbi_points_fp: str, osm_fp: str, logger: logging.Logger
) -> Tuple[Optional[QgsVectorLayer], Optional[QgsVectorLayer]]:
    """
    Load required layers with improved error handling.
    Tuple[Optional[QgsVectorLayer], Optional[QgsVectorLayer]]:
        A tuple containing the NBI points layer and the OSM ways layer.
    """

    def load_layer(fp: str, layer_name: str) -> Optional[QgsVectorLayer]:
        """
        Load a layer and log errors if the loading fails.
        """
        layer = QgsVectorLayer(fp, layer_name, "ogr")
        if not layer.isValid():
            logger.error(
                f"{layer_name} layer failed to load. Check the file path and ensure the file exists."
            )
            raise LayerLoadError(layer_name, "Failed to load layer.")
        return layer

    try:
        nbi_points_gl = load_layer(nbi_points_fp, "nbi-points")
        osm_gl = load_layer(osm_fp, "filtered")
    except LayerLoadError as e:
        logger.error(e)
        raise  # Re-raise the exception to be handled at a higher level

nitin-khandagale-mapup commented on hydrography-approach/processing_scripts/filter_data/filter_osm_ways.py:
```diff
@@ -50,4 +50,4 @@ def filter_ways(input_osm_pbf, output_osm_pbf, output_gpkg):
     convert_to_geopackage(output_osm_pbf, output_gpkg)
 
-    print(f"Output file: {output_gpkg} has been created successfully!")
+    logger.info(f"Output file: {output_gpkg} has been created successfully!")
```
Exceptions are not handled in the file. In other scripts also, exceptions were not handled.

varun-andhra-mapup replied:
I've added exception handling to the scripts.

nitin-khandagale-mapup commented on hydrography-approach/processing_scripts/associate_data/determine_final_osm_id.py:
```diff
@@ -199,13 +199,16 @@ def create_intermediate_association(df, intermediate_association):
+        df.to_csv(intermediate_association)
+        logger.info(f"{intermediate_association} file has been created successfully!")
+    except IOError as e:
+        logger.error(f"Failed to write {intermediate_association}: {str(e)}")
```
Handling exceptions for saving a CSV and logging is fine but other parts of the function should be handled as well.

nitin-khandagale-mapup replied:
Type hints are not given.

varun-andhra-mapup replied:
I've added exception handling and type hints to the scripts.

nitin-khandagale-mapup commented on hydrography-approach/processing_scripts/associate_data/determine_final_osm_id.py:
Object-oriented programming can help to make code easier to understand, easier to maintain, and easier to extend.

pushpak-gote-mapup commented on hydrography-approach/processing_scripts/filter_data/process_filter_nbi_bridges.py:
```diff
@@ -31,7 +31,7 @@ def exclude_duplicate_bridges(df, output_duplicate_exclude_csv):
 
 
-def convert_to_gpkg(df, output_gpkg_file):
+def convert_to_gpkg(df, output_gpkg_file, logger):
```
You can include exception handling and type hints here, also in other remaining functions as well

varun-andhra-mapup replied:
I've added exception handling and type hints to the scripts.

ajay-padwal-mapup commented:
Automatically closing this PR as it has been open for more than 60 days.
